### PR TITLE
Flexible Sync API

### PR DESCRIPTION
**Pull request by dianaafanador3**
This PR includes all the designed Flexible Sync APIs for Swift and Objective-C, doesn't include any implementation or testing even though it includes some new test files.
APIs are internal or on Private headers so we can do incremental PR review on this project.

### CODE REVIEW

File: Realm.podspec [hunk 1]
```diff
@@ -97,9 +97,12 @@ Pod::Spec.new do |s|
                               'include/RLMSet_Private.h',
                               'include/RLMSwiftProperty.h',
                               'include/RLMSyncConfiguration_Private.h',
+                              'include/RLMFlexibleSyncConfiguration_Private.h',
+                              'include/RLMFlexibleSyncConfiguration.h',
                               'include/RLMSyncUtil_Private.h',
                               'include/RLMUpdateResult_Private.h',
                               'include/RLMUser_Private.h',
+                              'include/RLMSyncSubscription.h',
```
Comment (leemaguire) [inline Realm.podspec:105]: I don't think it makes sense to reference these files in the podspec yet.

File: Realm.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -3257,13 +3306,30 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = 5D659E761BE03E0D006515A0 /* Realm.xcconfig */;
 			buildSettings = {
+				CLANG_ENABLE_MODULES = YES;
```
Comment (leemaguire) [inline Realm.xcodeproj/project.pbxproj:3309]: This looks like it should be omitted?

File: Realm.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -3371,6 +3437,7 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = 3FB56E7E250D457A00A6216B /* ObjectServerTests.xcconfig */;
 			buildSettings = {
+				ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES = YES;
```
Comment (leemaguire) [inline Realm.xcodeproj/project.pbxproj:3440]: This also looks like it should be omitted.

File: Realm/RLMFlexibleSyncConfiguration_Private.hpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import "RLMFlexibleSyncConfiguration_Private.h"
+
+@interface RLMFlexibleSyncConfiguration ()
+
+- (instancetype)initWithUser:(RLMUser *)user;
```
Comment (leemaguire) [inline Realm/RLMFlexibleSyncConfiguration_Private.hpp:23]: This is duplicated in the corresponding `.h` file.

File: Realm/RLMSyncSubscription_Private.h [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark - Subscription States
+
+typedef NS_ENUM(NSUInteger, RLMSyncSubscriptionState) {
+    RLMSubscriptionStateComplete,
+
```
Comment (leemaguire) [inline Realm/RLMSyncSubscription_Private.h:27]: Even though this is private at the moment, it would not be a bad idea to start adding some docstrings.

File: RealmSwift/SyncSubscription.swift [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    // collection of subscription
+    @discardableResult
+    internal func write(_ block: (() throws -> Void)) async throws -> SyncSubscriptionTask {
+        fatalError()
+    }
+}
+#endif // swift(>=5.5)
+
+#if !(os(iOS) && (arch(i386) || arch(arm)))
+import Combine
```
Comment (jsflax) [inline RealmSwift/SyncSubscription.swift:201]: Combine stuff should probably be a part of a separate PR.

File: RealmSwift/SyncSubscription.swift [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    }
+}
+
+#if swift(>=5.5) && canImport(_Concurrency)
+@available(macOS 12.0, tvOS 15.0, iOS 15.0, watchOS 8.0, *)
+extension SyncSubscriptionTask {
+    // Notifies state changes for the write subscription transaction
+    // if state is complete this will return complete, will
+    // throw an error if someone updates the subscription set while waiting
+    internal func observe() -> AsyncStream<SyncSubscriptionState> {
```
Comment (jsflax) [inline RealmSwift/SyncSubscription.swift:110]: If we're using an AsyncStream here, this should be a computed variable named `state`.

File: RealmSwift/SyncSubscription.swift [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+    // When the subscription was last updated. Recorded automatically.
+    internal var updatedAt: Date = Date()
+
+    // Name of the subscription, if not specified it will return the value in Query
+    internal var name: String = ""
+
+    #if swift(>=5.4)
+    // Update query for subscription
+    internal func update(@QueryBuilder _ to: () -> (AnySyncSubscription)) throws {
```
Comment (jsflax) [inline RealmSwift/SyncSubscription.swift:55]: We are missing a way to update queries for swift <5.4.

Comment (dianaafanador3) [reply RealmSwift/SyncSubscription.swift:55]: I'm adding this for now because we are supporting Xcode 12.2 but referencing this PR https://github.com/realm/realm-cocoa/pull/7501 this is something that is going to change in the near future, do we want to add another API?

File: Realm/RLMSyncSubscription_Private.h (hunk 5 above)
Comment (dianaafanador3) [reply Realm/RLMSyncSubscription_Private.h:27]: Added some docstrings for the Swift API, will add objective-C later

File: RealmSwift/SyncSubscription.swift [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+extension SyncSubscriptionState: Equatable {
+    static func == (lhs: SyncSubscriptionState, rhs: SyncSubscriptionState) -> Bool {
+        true
+    }
+}
+
+internal protocol AnySyncSubscription {}
+
+internal class SyncSubscription<Element: Object>: AnySyncSubscription {
```
[the remark this answers is not in the document]
Comment (dianaafanador3) [reply RealmSwift/SyncSubscription.swift]: I changes this to use a type of type-erasure, this is still a WIP

File: RealmSwift/Sync.swift [hunk 10]
```diff
@@ -787,3 +787,166 @@ extension FunctionCallable {
...
+extension Realm {
+    // Get all subscriptions for this Realm.
+    /*private(set)*/ public var subscriptions: [AnySubscription] {
+        fatalError()
+    }
+}
+
+// TODO: Can we observer changes on the subscription set?
+// Task to get state changes from the write transaction
+@frozen public struct SubscriptionTask {
```
[the remark this answers is not in the document]
Comment (dianaafanador3) [reply RealmSwift/Sync.swift]: This will allow us do this, we can think in another approach for this later
       ```
 try subscriptions.write {
            subscriptions.append {
                SyncSubscription<Contact>(name: "contacts-ny") {
                    $0.address.state == "NY" && $0.age > 10
                }
            }
        }
        .observe { state in
            if case .complete = state {

            }
        }
```
